mrmustard/lab_dev/wires.py

```diff
@@ -0,0 +1,193 @@
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Classes for supporting tensor network functionalities."""
+
+from __future__ import annotations
+
+from typing import Optional, Sequence, Union
+import uuid
+
+from ..utils.typing import Mode
+
+
+__all__ = ["Wire", "Wires"]
+
+Wire = int
+r"""
+An integer representing a wire in a tensor network.
+"""
+
+
+# pylint: disable=too-many-boolean-expressions
+class Wires:
+    r"""
+    A class with wire functionality for tensor network tensor applications.
+
+    In general we distinguish between input and output wires, and between ket and bra sides.
+
+    Args:
+        modes_out_bra: The output modes on the bra side.
+        modes_in_bra: The input modes on the bra side.
+        modes_out_ket: The output modes on the ket side.
+        modes_in_ket: The input modes on the ket side.
+
+    .. jupyter-execute::
+
+        from mrmustard.lab_dev.wires import Wires
+
+        # initialize modes
+        modes = [0, 1]
+
+        # initialize `wires` object with given modes
+        wires = Wires(modes_out_ket = modes, modes_in_ket = modes)
+    """
+
+    def __init__(
+        self,
+        modes_out_bra: Optional[Sequence[Mode]] = None,
+        modes_in_bra: Optional[Sequence[Mode]] = None,
+        modes_out_ket: Optional[Sequence[Mode]] = None,
+        modes_in_ket: Optional[Sequence[Mode]] = None,
+    ) -> None:
+        modes_out_bra = modes_out_bra or []
+        modes_in_bra = modes_in_bra or []
+        modes_out_ket = modes_out_ket or []
+        modes_in_ket = modes_in_ket or []
+
+        modes = modes_out_bra or modes_in_bra or modes_out_ket or modes_in_ket
+        if (
+            (modes_out_bra and modes_out_bra != modes)
+            or (modes_in_bra and modes_in_bra != modes)
+            or (modes_out_ket and modes_out_ket != modes)
+            or (modes_in_ket and modes_in_ket != modes)
+        ):
+            # cannot define the list of modes unambiguously
+            self._modes = None
+        else:
+            self._modes = list(modes)
+
+        keys = self._modes or set(modes_in_ket + modes_out_ket + modes_in_bra + modes_out_bra)
+        self._out_bra = {m: uuid.uuid4().int if m in modes_out_bra else None for m in keys}
+        self._in_bra = {m: uuid.uuid4().int if m in modes_in_bra else None for m in keys}
+        self._out_ket = {m: uuid.uuid4().int if m in modes_out_ket else None for m in keys}
+        self._in_ket = {m: uuid.uuid4().int if m in modes_in_ket else None for m in keys}
+
+    @property
+    def in_bra(self) -> dict[Mode, Optional[Wire]]:
```

Comment:
didn't we agree on reintroducing the `.output`, `.input`, `.ket` and `.bra` properties and return a new `Wires` object?

Reply:
that will save you from having to define all the combinations as special methods (`modes_out_bra`, `modes_in_bra` etc, as well as `out_bra`, `out_ket` etc)

Reply:
that will also allow you to get more general subsets (e.g. all the output wires regardless of ket/bra/mode)

Reply:
Last time we talked about that, we agreed on introducing only what we strictly need. 

I need to be able to **read** the dictionaries `_in_bra`, `_out_bra` etc., or the modes that go in and out, which is what these properties allow you to do. I don't need to generate new wires every time

Reply:
P.S. By "I don't need to generate new wires every time" I mean that I would still need these properties, in order to be able to read the dictionaries and the modes